Support drag and drop in TaurusPlot

cmft opened the pull request:
Implement drag and drop for TaurusPlot

cmft committed:
Refactor TaurusXYModelChooserTool._onTriggered

Refactor part of TaurusXYModelChooserTool._onTriggered method code
in several method: getTaurusPlotDataItemsConf and updatePlotItems.

This method has been refactor to do the code more usable.

cmft committed:
Support drag and drop in TaurusPlot

The drag and drop suppor is delegarted in TaurusXYModelChooserTool.
In TaurusXYModelChooserTool class is implemented a eventFilter method
for its treat the parent events and dropMimeData method to get the
dropped models.

cmft committed:
Remove duplicated code

updatePlotItems has unnecessary code due to a merge problem

cpascual commented on taurus_pyqtgraph/taurusmodelchoosertool.py:
```diff
@@ -29,14 +29,17 @@
 __all__ = ["TaurusModelChooserTool", "TaurusImgModelChooserTool"]
 
-from taurus.external.qt import QtGui
+from taurus.external.qt import Qt, QtGui
```
Since you are importing Qt, importing QtGui seems redundant. I suggest to replace QtGui usages by Qt

cpascual commented on taurus_pyqtgraph/taurusmodelchoosertool.py:
```diff
@@ -202,6 +205,10 @@ def __init__(self, parent=None):
+        self._parent = parent
+        if parent is not None:
+            # Filter drag and drop parent's event
+            parent.installEventFilter(self)
```
For a tpg tool, the parent should not be checked or used in the constructor. Instead it should be delayed until  `attachToPlotItem` or `_onTriggered`.

cmft replied:
> For a tpg tool, the parent should not be checked or used in the constructor. Instead it should be delayed until attachToPlotItem or _onTriggered.

I moved it to setParent method.

cmft committed:
Replace QtGui usages by Qt

cmft committed:
Move add event filter to setParent method

Reimplement setParent to add a event filter

cmft committed:
Current models are deleting after a drop model

The drop event deletes the existing plot data items conf. models.

Fix it, getting the existing one before ading the new ones.

cpascual commented on taurus_pyqtgraph/taurusmodelchoosertool.py:
```diff
@@ -264,7 +264,7 @@ def dropMimeData(self, data):
 
         if len(models) > 0:
-            conf_list = []
+            conf_list = self.getTaurusPlotDataItemsConf()
```
just a reminder for looking at this carefully: we need yet to define the expected behaviour when setting/ adding models in tpg.

Possibilities:
- a) replace all existing models (this is what `...tpg.TaurusPlot.setModel` and the previous behavior of the drop do
- b) add models unconditionally (this is what would be done by a drop after this change). This would duplicate curves if they already exist
- c) add models only if they are not already added (this is what `...qwt5.TaurusPlot.addModel` or a drop in the old qwt5 widgets do)

I, personally think that the best is to implement addModel with strategy "b" in the tpg's modelchooser, and delegate to it in `tpg.TaurusPlot.addModel` as well as for the drops.

Commit:
Merge branch 'supportdragdrop' of https://github.com/cmft/taurus_pyqtgraph into cmft-supportdragdrop

Fixed conflicts in:
	taurus_pyqtgraph/taurusmodelchoosertool.py

Commit:
Avoid keeping custom reference to parent

Use QObject.parent() instead of keeping our own reference

Commit:
Fix indentation problems caused by last merge

Last merge introduced some wrong indentatios. Fix them

Commit:
(m) PEP8

Commit:
Merge branch 'master' into cmft-supportdragdrop

Commit:
Merge branch 'master' into cmft-supportdragdrop

Commit:
[py3] Fix decoding issue when dropping models on a tpg plot

The dropMimeData implementation in tpg model chooser does not properly
decode the mimedata. Fix it to make it work with py3.

Commit:
Refactor TaurusXYModelChooserTool

Refactor the logic and API of TaurusXYModelChooserTool to:
- simplify drag & drop implementation
- add getModelNames, addModels methods
- support z-ordering
- make some of the previous API private
- avoid hardcoding itemClass

cpascual commented:
IMHO the tool is now ready.

lt can be tested with a tpg.TaurusPlot:
```console 
python -m taurus.qt.qtgui.tpg.plot
```

or with a pure PlotWidget:
```python
from taurus.qt.qtgui.application import TaurusApplication
from taurus.qt.qtgui.tpg.taurusmodelchoosertool import TaurusXYModelChooserTool
import pyqtgraph as pg

if __name__ == '__main__':
    import sys
    app = TaurusApplication()

    w = pg.PlotWidget()

    t = TaurusXYModelChooserTool()
    t.attachToPlotItem(w.getPlotItem(), parentWidget=w)

    w.show()
    ret = app.exec_()
    sys.exit(ret)
```
